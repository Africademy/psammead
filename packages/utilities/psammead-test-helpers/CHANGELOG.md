# Psammead Test Helpers Changelog

| Version | Description |
|---------|-------------|
| 0.1.2   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
