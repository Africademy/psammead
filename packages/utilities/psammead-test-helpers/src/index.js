import { render, waitForDomChange } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';

const renderWithHelmet = async component => {
  /*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:
 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });
 */

  const { container } = render(component);
  const noop = () => {};
  const ARBITRARY_TIMEOUT = 100; // seems long enough for any dom mutations to occur

  return waitForDomChange({
    container: document.querySelector('head'),
    timeout: ARBITRARY_TIMEOUT,
  })
    .catch(noop) // handle a waitForDomChange timeout
    .then(mutationsList => {
      const domMutationDetected = mutationsList && mutationsList.length;

      if (domMutationDetected) {
        // helmet was probably used so we should get the full html

        const htmlElement = document.querySelector('html');

        const helmetElements = document.querySelectorAll(
          '[data-react-helmet="true"]',
        );

        const removeHelmetAttributes = el =>
          el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

        removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

        Array.from(helmetElements).forEach(removeHelmetAttributes);

        return { container: document.querySelector('html') };
      }

      return { container };
    });
};

export const shouldMatchSnapshot = (title, component) => {
  it(title, async () => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    const { container } = await renderWithHelmet(component);
    const hasOneChild = container.children.length === 1;
    /*
     * if the container has more than one child then it's a component that uses a
     * fragment at the top level so we should not select the first child because it
     * wouldn't snapshot the whole component
     */
    expect(
      hasOneChild ? removeWrappingDiv(container) : container,
    ).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
