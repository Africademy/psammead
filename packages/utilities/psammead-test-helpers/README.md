# psammead-test-helpers &middot; [![GitHub license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/BBC-News/psammead/blob/latest/LICENSE) [![npm version](https://img.shields.io/npm/v/@bbc/psammead-test-helpers.svg)](https://www.npmjs.com/package/@bbc/psammead-test-helpers) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://reactjs.org/docs/how-to-contribute.html#your-first-pull-request)

This package is a collection helper methods used for snapshot testing, with support for styled components.

## Installation

```jsx
npm install @bbc/psammead-test-helpers --save
```

## Usage

```jsx
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';

shouldMatchSnapshot(
  'should render correctly',
  <h1>Hello World</h1>,
);
```

## Roadmap

There is currently a plan to migrate from `react-test-renderer` to `enzyme` as it provides a nicer higher-level API over the same functionality, as well as having support for more granular unit tests.

## Contributing

Psammead is completely open source. We are grateful for any contributions, whether they be new components, bug fixes or general improvements. Please see our primary contributing guide which can be found at [the root of the Psammead respository](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md).

### [Code of Conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md)

We welcome feedback and help on this work. By participating in this project, you agree to abide by the [code of conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md). Please take a moment to read it.

### License

Psammead is [Apache 2.0 licensed](https://github.com/BBC-News/psammead/blob/latest/LICENSE).
