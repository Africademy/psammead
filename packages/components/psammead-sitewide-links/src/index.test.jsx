import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import SitewideLinks from './index';

describe(`SitewideLinks`, () => {
  const link = {
    href: 'https://www.bbc.co.uk/news',
    text: 'Link',
  };
  const ccpaLink = {
    href:
      'https://www.bbc.com/usingthebbc/cookies/how-can-i-change-my-bbc-cookie-settings/',
    text: 'AdChoices / Do Not Sell My Info',
    lang: 'en-GB',
  };

  const links = new Array(7).fill(link).map((linkItem, index) => ({
    ...linkItem,
    text: `${linkItem.text}${index}`,
  }));

  const ccpaLinks = new Array(8).fill(link, 0, 7).map((linkItem, index) => ({
    ...linkItem,
    text: `${linkItem.text}${index}`,
  }));

  ccpaLinks[7] = ccpaLink;

  shouldMatchSnapshot(
    'should render correctly',
    <SitewideLinks
      links={links}
      copyrightText={<span>Text here.</span>}
      externalLink={link}
      service="news"
    />,
  );

  shouldMatchSnapshot(
    'should render correctly with CCPA link',
    <SitewideLinks
      links={ccpaLinks}
      copyrightText={<span>Text here.</span>}
      externalLink={link}
      service="news"
    />,
  );

  shouldMatchSnapshot(
    'should render correctly with trust project link',
    <SitewideLinks
      links={links}
      trustProjectLink={{
        href: 'http://www.bbc.com/terms',
        text: 'Why you can trust the bbc',
      }}
      copyrightText={<span>Text here.</span>}
      externalLink={link}
      service="news"
    />,
  );
});
