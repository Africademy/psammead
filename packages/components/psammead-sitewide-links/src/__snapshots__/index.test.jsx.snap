// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SitewideLinks should render correctly 1`] = `
<div
  class="SitewideLinksWrapper-sc-1gggobt-0 fAdZsy"
>
  <div
    class="ConstrainedWrapper-sc-1gggobt-1 gZsHvP"
  >
    <ul
      class="StyledList-sc-2e1ekc-0 Hkot"
      role="list"
    >
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link0
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link1
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link2
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link3
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link4
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link5
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link6
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.com/usingthebbc/cookies/how-can-i-change-my-bbc-cookie-settings/"
          lang="en-GB"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            AdChoices / Do Not Sell My Info
          </span>
        </a>
      </li>
    </ul>
    <p
      class="StyledParagraph-sc-1gggobt-2 jQPFnr"
    >
      <span>
        Text here.
      </span>
       
      <a
        class="StyledLink-nhmg8j-0 jHQlUp"
        href="https://www.bbc.co.uk/news"
      >
        <span
          class="StyleLinkText-nhmg8j-1 fPnERB"
        >
          Link
        </span>
      </a>
    </p>
  </div>
</div>
`;

exports[`SitewideLinks should render correctly with trust project link 1`] = `
<div
  class="SitewideLinksWrapper-sc-1gggobt-0 fAdZsy"
>
  <div
    class="ConstrainedWrapper-sc-1gggobt-1 bpERjU"
  >
    <ul
      class="StyledList-sc-2e1ekc-0 lhGfpo"
      role="list"
    >
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="http://www.bbc.com/terms"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Why you can trust the bbc
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link0
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link1
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link2
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link3
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link4
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link5
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.co.uk/news"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            Link6
          </span>
        </a>
      </li>
      <li
        class="StyledListItem-sc-2e1ekc-1 kUOQdM"
        role="listitem"
      >
        <a
          class="StyledLink-nhmg8j-0 dHbSzF"
          href="https://www.bbc.com/usingthebbc/cookies/how-can-i-change-my-bbc-cookie-settings/"
          lang="en-GB"
        >
          <span
            class="StyleLinkText-nhmg8j-1 fPnERB"
          >
            AdChoices / Do Not Sell My Info
          </span>
        </a>
      </li>
    </ul>
    <p
      class="StyledParagraph-sc-1gggobt-2 jQPFnr"
    >
      <span>
        Text here.
      </span>
       
      <a
        class="StyledLink-nhmg8j-0 jHQlUp"
        href="https://www.bbc.co.uk/news"
      >
        <span
          class="StyleLinkText-nhmg8j-1 fPnERB"
        >
          Link
        </span>
      </a>
    </p>
  </div>
</div>
`;
